core-bundle/src/Security/Authenticator/ContaoLoginAuthenticator.php

```diff
@@ -0,0 +1,232 @@
+        }
+
+        $this->framework->initialize();
+
+        $page = $request->attributes->get('pageModel');
+
+        if (!$page instanceof PageModel) {
+            throw new \RuntimeException('No PageModel found on request.');
+        }
+
+        $page->loadDetails();
+        $page->protected = false;
+
+        $route = $this->pageRegistry->getRoute($page);
+        $subRequest = $request->duplicate(null, null, $route->getDefaults());
+
+        if (null === $this->tokenStorage->getToken()) {
+            $pageAdapter = $this->framework->getAdapter(PageModel::class);
+            $errorPage = $pageAdapter->findFirstPublishedByTypeAndPid('error_401', $page->rootId);
+
+            if (null === $errorPage) {
+                throw new PageNotFoundException('No error page found.');
+            }
+
+            $errorPage->loadDetails();
+            $errorPage->protected = false;
+
+            $route = $this->pageRegistry->getRoute($errorPage);
+            $subRequest = $request->duplicate(null, null, $route->getDefaults());
+        }
+
+        try {
+            return $this->httpKernel->handle($subRequest, HttpKernelInterface::SUB_REQUEST, false);
+        } catch (ResponseException $e) {
+            return $e->getResponse();
+        }
+    }
+
+    public function supports(Request $request): ?bool
+    {
+        return $request->isMethod('POST')
+            && $request->request->has('FORM_SUBMIT')
+            && preg_match('/^tl_login(_\d+)?$/', (string) $request->request->get('FORM_SUBMIT'));
+    }
+
+    public function authenticate(Request $request): Passport
+    {
+        // When the firewall is lazy, the token is not initialized in the "supports" stage, so this check does only work
+        // within the "authenticate" stage.
+        $currentToken = $this->tokenStorage->getToken();
+
+        if ($currentToken instanceof TwoFactorTokenInterface) {
+            return $this->twoFactorAuthenticator->authenticate($request);
+        }
+
+        $credentials = $this->getCredentials($request);
+
+        $passport = new Passport(
+            new UserBadge($credentials['username'], [$this->userProvider, 'loadUserByIdentifier']),
+            new PasswordCredentials($credentials['password']),
+            [new RememberMeBadge()]
+        );
+
+        if ($this->options['enable_csrf']) {
+            $passport->addBadge(new CsrfTokenBadge($this->options['csrf_token_id'], $credentials['csrf_token']));
+        }
+
+        if ($this->userProvider instanceof PasswordUpgraderInterface) {
+            $passport->addBadge(new PasswordUpgradeBadge($credentials['password'], $this->userProvider));
+        }
+
+        return $passport;
+    }
+
+    public function createToken(Passport $passport, string $firewallName): TokenInterface
+    {
+        $credentialsBadge = $passport->getBadge(TwoFactorCodeCredentials::class);
+
+        if ($credentialsBadge instanceof TwoFactorCodeCredentials) {
```

Comment:
might want to early return here instead with parent::createToken

Reply:
Changed in 68fc2389dd.